### DISCUSSION TITLE
ClassCastException in Excel emitter

### DISCUSSION THREAD

**Discussion by jochenberger** [eclipse-birt/birt#946] [General]
Hi there,

I'm getting the following exception when trying to create an XLSX report. The report contains an HTML table, maybe that's a problem.
I'm using Birt 4.9 BTW.

```
...
Caused by: java.lang.ClassCastException: class org.eclipse.birt.report.engine.content.impl.ContainerContent cannot be cast to class org.eclipse.birt.report.engine.content.impl.CellContent (org.eclipse.birt.report.engine.content.impl.ContainerContent and org.eclipse.birt.report.engine.content.impl.CellContent are in unnamed module of loader org.eclipse.osgi.internal.loader.EquinoxClassLoader @31f04847)
	at uk.co.spudsoft.birt.emitters.excel.handlers.AbstractRealTableCellHandler.startTable(AbstractRealTableCellHandler.java:188)
	at uk.co.spudsoft.birt.emitters.excel.ExcelEmitter.startTable(ExcelEmitter.java:297)
```

**Comment by claesrosell** [comment]
Can you provide a report design with which one can reproduce the error?

**Reply by jochenberger** [reply]
The report file is https://github.com/SerNet/verinice/blob/1.23.1/sernet.gs.server/WebContent/WEB-INF/reportDeposit/bp_riskmatrix.rptdesign.

**Comment by claesrosell** [comment]
I converted this to an issue

**Reply by jochenberger** [reply]
Thanks. It looks as if it turns out to be a non-issue though.

**Reply by claesrosell** [reply]
From my point of view, a ClassCastException is a bug even if HTML tables are not supported. The creation of the Excel document should not fail even if the table is not correctly rendered. I would like to investigate it, but I will not put oceans of time into it.

**Comment by wimjongman** [comment]
Claes, there is a function for this:

![image](https://user-images.githubusercontent.com/180969/163994020-ebe783f9-137f-44d0-96ac-58bc8e6ea395.png)

**Comment by hvbtup** [comment]
HTML tables are not supported.

**Comment by hvbtup** [comment]
They "work" in a very very very limited way in the PDF emitter and probably in the HTML emitter, but they are not supported. Only a very limited subset of old-fashioned HTML is supported, which allows choosing font properties and color.

**Reply by jochenberger** [reply]
Oh, that explains it I guess. DOC and DOCX seem to work too, so I assumed it is a problem with the Excel emitter.
Maybe the error reporting could be improved here? :slightly_smiling_face: 
Btw.: The report file is https://github.com/SerNet/verinice/blob/1.23.1/sernet.gs.server/WebContent/WEB-INF/reportDeposit/bp_riskmatrix.rptdesign

**Reply by claesrosell** [reply]
I put some effort into this but the ClassCastException is not reproducible with the referenced report design, the report design seems to use a custom data source.
Is it possible to provide a .rptdesign which reproduces the problem and uses the "example database"?

**Reply by jochenberger** [reply]
I'm not familiar with report creation I'm afraid.
But I guess I can help with debugging.
The problem seems to be the cast in https://github.com/eclipse/birt/blob/birt_4_9_0/engine/uk.co.spudsoft.birt.emitters.excel/src/uk/co/spudsoft/birt/emitters/excel/handlers/AbstractRealTableCellHandler.java#L188.
We're casting `table.getParent()` to a `CellContent` and that does not work because the parent is a `ContainerContent`. But that cast might be to strict anyway. If we only want to be able to call `getGenerateBy`, then we can cast to `AbstractContent` instead which is a common superclass to both `CellContent` and `ContainerContent`.
I wonder why there is an additional type check in https://github.com/eclipse/birt/blob/birt_4_9_0/engine/uk.co.spudsoft.birt.emitters.excel/src/uk/co/spudsoft/birt/emitters/excel/handlers/AbstractRealTableCellHandler.java#L198 though. Unfortunately, that's where my understanding of the emitter code ends. :laughing:

**Reply by jochenberger** [reply]
> If we only want to be able to call getGenerateBy, then we can cast to AbstractContent instead [...]

Oh, no, if we did that, we'd get a NPE because `table.parent.generateBy` is `null` in my case.
So maybe there should also be type checks in that branch?

**Reply by jochenberger** [reply]
The XLS emitter works btw.

**Reply by claesrosell** [reply]
I have put a couple of more hours into this issue, trying to recreate the data set with fake data. Unfortunately I have not been able to reproduce the error, probably because the result set should include json that is later used by javascript to create some parts of the report.

I am willing to do a test and alter the code on Line 188. -Just do a cast to org.eclipse.birt.report.engine.content.IContent instead (since the called method is defined in that interface). It should not affect the emitter in a negative way, the failure may still exist but the effects of it could be different.
@jochenberger , if I do a PR for this, could you test it?

**Reply by jochenberger** [reply]
@claesrosell 
I'll happily do that but mind my comment https://github.com/eclipse/birt/discussions/946#discussioncomment-2606416. I think that it'll only cause an NPE instead of a ClassCastException.

**Reply by jochenberger** [reply]
Oh, I could provide some JSON for the report if that would help you.

**Reply by claesrosell** [reply]
Any help to reproduce the problem is welcome. A .rptdocument might work as well. Not sure though,.

**Reply by UrsZeidler** [reply]
> I put some effort into this but the ClassCastException is not reproducible with the referenced report design, the report design seems to use a custom data source. Is it possible to provide a .rptdesign which reproduces the problem and uses the "example database"?

> Any help to reproduce the problem is welcome. A .rptdocument might work as well. Not sure though,.

Here is such a file, you will need to rename to back to *.rptdesign, it contains no datasource no data set.

Should throw a 

`Caused by: java.lang.ClassCastException: class org.eclipse.birt.report.engine.content.impl.ContainerContent cannot be cast to class org.eclipse.birt.report.engine.content.impl.CellContent (org.eclipse.birt.report.engine.content.impl.ContainerContent and org.eclipse.birt.report.engine.content.impl.CellContent are in unnamed module of loader org.eclipse.osgi.internal.loader.EquinoxClassLoader @5c9ad872)
	at uk.co.spudsoft.birt.emitters.excel.handlers.AbstractRealTableCellHandler.startTable(AbstractRealTableCellHandler.java:188)
`
when rendered in XLSX.
The pdf and html are working fine.

[bp_riskmatrix.txt](https://github.com/eclipse/birt/files/8590427/bp_riskmatrix.txt)
